device: Improve Allied Telesis support within LibreNMS

mattie47 opened the pull request:
As per #7652, have tried to update Allied Telesis support within LibreNMS.

This is currently partially working, with assistance requested on improving  alliedtelesis.inc.php

**Work done:**
- Removed very old Allied Telesis mibs.
- Removed "allied*" files, and replaced with more correct name.
- Added latest available version of Allied Telesis mibs (Source is me as I'm a test engineer at Allied Telesis)
- Added SVGs os and logo. (Source is me as I'm a test engineer at Allied Telesis)
-- Optimized SVGs as per the docs. 
- moved allied.inc.php to alliedtelesis.inc.php
-- Updated mibs
-- This file still needs some work as I don't know PHP but have done what I can...
- Fixed issues based on ./scripts/pre-commit.php


I'm surprised people were even using the allied.inc.php in its current state. None of the mibs work correctly for Allied Telesis products running AW+ (The primary OS of almost all switches and routers from the last 7 years). Doing an snmpget for the original 3 variables (atiswitchProductType.0, atiswitchSwVersion.0, atiswitchSw.0 returns nothing).



DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/librenms/librenms?pullRequest=7670) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/librenms/librenms?pullRequest=7670) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: mattie47<br/>:x: Matt Read<br/><hr/>**Matt Read** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/librenms/librenms?pullRequest=7670) it.</sub>

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/0bd75aa4-6fba-41c8-8768-ef5add28f2c7) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/0bd75aa4-6fba-41c8-8768-ef5add28f2c7) completed: **No new issues**

mattie47 commented:
I say this "partially" works for the following reasons:

- I changed the snmpget to snmpgetnext for $hardware and $serial, which seemed to work initially:

![image](https://user-images.githubusercontent.com/4696925/32522862-dbf31a4a-c47d-11e7-8fa1-462352d8035f.png)


However the polling seems to be broken as a result:

![image](https://user-images.githubusercontent.com/4696925/32522923-12ff1a0c-c47e-11e7-91f8-49a8c4059632.png)


And I can see it's causing these issues: 

https://pastebin.com/tNShV9Kg

Why am I using snmpgetnext? Because the mib oid is not consistent and I'm not sure how else to do it.

As an example:


```
librenms@librenms:~$  snmpwalk -OUsb -v3 -l authPriv -u librenms -a SHA -A "librenms" -x AES -X "librenms"  -t 30 10.37.105.3   rscBoardName
rscBoardName.1.1 = STRING: AT-x550-18XTQ
librenms@librenms:~$  snmpwalk -OUsb -v3 -l authPriv -u librenms -a SHA -A "librenms" -x AES -X "librenms"  -t 30 10.37.105.4   rscBoardName
rscBoardName.1.3 = STRING: AT-x930-28GTX
rscBoardName.1.4 = STRING: AT-StackQS <----- stacking module
rscBoardName.1.5 = STRING: PWR250 <------ Power supply
librenms@librenms:~$  snmpwalk -OUsb -v3 -l authPriv -u librenms -a SHA -A "librenms" -x AES -X "librenms"  -t 30 10.37.105.1   rscBoardName
rscBoardName.1.3 = STRING: AT-SBx908 GEN2 <------ Device model
rscBoardName.1.4 = STRING: AT-XEM2-12XS <---- Expansion module
rscBoardName.1.5 = STRING: AT-XEM2-4QS
rscBoardName.1.6 = STRING: AT-SBxPWRSYS2/AC
rscBoardName.1.7 = STRING: AT-FAN08
rscBoardName.1.8 = STRING: AT-FAN08
rscBoardName.2.3 = STRING: AT-SBx908 GEN2
rscBoardName.2.4 = STRING: AT-XEM2-12XS
rscBoardName.2.5 = STRING: AT-XEM2-4QS
rscBoardName.2.6 = STRING: AT-SBxPWRSYS2/AC
rscBoardName.2.7 = STRING: AT-FAN08
rscBoardName.2.8 = STRING: AT-FAN08
```

The first returned boardname always seems to be the primary unit (i.e. baseboard unit and not one of the expansion modules). 

I've tested this on a number of different model products, however once I've had some help with this config file, I can test on some more obscure Allied Telesis hardware...


Thanks,

Matt

murrant commented:
Thanks for working to improve the support for your devices :)

1. Your logo needs to be minimized: https://jakearchibald.github.io/svgomg/
2. If they are duplicate, you only need an svg in image/os, image/logos for logos that don't work well as small icon.
3. There is no snmp_getnext() function at this time :)  We wouldn't be opposed to getting that implemented.
4. We'll need to send a notification out about the os rename.  In cause users have alerts or groups based on the os name.  See: https://github.com/librenms/librenms.github.io/pull/96

Because you might know, do you have any idea why we have a radlan os?  It looks the same to me but slightly different.  A third OS by Allied Telesis?

With the old OS, users would still get port info, etc.  Just maybe not correct hardware/os info. As a fun FYI, the top 10 from our user contributed stats:
```
|Total|OS    |sysObjectID             | sysDescr |
|   25|allied|.1.3.6.1.4.1.207.1.4.167| AT-GS950/24 Gigabit Ethernet WebSmart Switch |
|   25|allied|.1.3.6.1.4.1.207.1.4.127| ATI AT-8000S |
|   24|allied|.1.3.6.1.4.1.207.1.4.128| ATI AT-8000S |
|   13|allied|.1.3.6.1.4.1.207.1.4.145| 48-port 10/100/1000 Ethernet Switch |
|   12|allied|.1.3.6.1.4.1.207.1.4.144| 24-port 10/100/1000 Ethernet Switch with PoE |
|    9|allied|.1.3.6.1.4.1.207.1.4.198| Allied Telesis router/switch, AW  v5.4.5-2.2 |
|    8|allied|.1.3.6.1.4.1.207.1.4.237| AlliedWare Plus (TM) 2.4.1.0 |
|    4|allied|.1.3.6.1.4.1.207.1.14.53| Allied Telesyn AT-9924T-EMC version 2.6.6-01 15-Apr-2005 |
|    3|allied|.1.3.6.1.4.1.207.1.14.49| Allied Telesis AT-9924T version 2.9.2-14 24-Jun-2014 |
|    3|allied|.1.3.6.1.4.1.207.1.4.151| AlliedWare Plus (TM) 2.1.8.0 |
```

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/067e2070-bbff-4933-8ac9-84f24413647f) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/067e2070-bbff-4933-8ac9-84f24413647f) completed: **No new issues**

mattie47 commented:
Hi @murrant 

Thanks for the feedback.

1. Done.
2. Have removed the duplicate.
3. I see you've done this in #7678. Thanks!
4. Yip No problem. I want to ensure what's in this PR works for everyone before the PR gets merged anyway. 
5. Yeah - I might ask someone at work today what to do here, otherwise I'll ask on Discord. I'm just not sure how to undo 06ea936 as I originally did my first commit from a local account and not from my github account. 

Thanks for providing the user contributed stats. I can probably find a GS950 and 8000S and one or two others to confirm they still work fine. 

>Because you might know, do you have any idea why we have a radlan os? It looks the same to me but slightly different. A third OS by Allied Telesis?

**Allied Telesyn/Telesis had/have about 3 main operating systems:**
- Alliedware - The original premium OS for Routers and Switches. This was a monolithic OS and any new development stopped I think around 2010, with the last bugfix in 2.9.2 in 2014.
- Alliedware Plus - The current and primary OS. First release was in 2007. This is what all routers and switches (minus some websmart devices) run.
- AlliedWare Plus version 2.x.x and older - Annoyingly this is not the same as Alliedware Plus 5.x.x. While the CLI is similiar it's completely different under the hood. This is found only on very low end platforms, and in recent years most low end platforms are now AW+ 5.x.x. 

Regarding radlan - This is an OS developed by Marvell and seems to incorporate a number of rebranded low-end devices: http://www.snmplink.org/cgi-bin/nd/m/Ent/R/RAD%20Network%20Devices%20(RND)/Bridges*A6Routers/RADLAN%20OS%20(ROS)/

I see 2 EOL low end Allied Telesis switches on there (ATI 8000S and 8000GS).

I'm still seeing a few issues with my new alliedtelesis.inc.php config which I'm currently investigating.


Thanks,

Matt

laf commented:
> Yeah - I might ask someone at work today what to do here, otherwise I'll ask on Discord. I'm just not sure how to undo 06ea936 as I originally did my first commit from a local account and not from my github account.

You can just squash all your commits in to one: https://github.com/todotxt/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit

mattie47 commented:
@murrant  @laf 

I've been banging my head against a wall trying to figure out why all my devices are now coming up as generic. As per below OID .1.3.6.1.4.1.12148.10.2.6.0 is failing, but I'm not sure where it's getting this from....

```
librenms@librenms:~$ ./discovery.php -h c3-vpn -d
LibreNMS Discovery
===================================
Version info:
Commit SHA: b898bdde1f90598ef998cc7786e298d6ca58aa7c
Commit Date: 1510176252
DB Schema: 215
PHP: 7.0.22-0ubuntu0.16.04.1
MySQL: 10.0.31-MariaDB-0ubuntu0.16.04.2
RRDTool: 1.5.5
SNMP: NET-SNMP 5.7.3
==================================DEBUG!
SQL[SELECT * FROM `devices` WHERE status = 1 AND disabled = 0 AND `hostname` LIKE 'c3-vpn' ORDER BY device_id DESC] 
SQL[SELECT * FROM devices_attribs WHERE `device_id` = '13'] 
c3-vpn 13 generic SQL[INSERT INTO `device_perf` (`xmt`,`rcv`,`loss`,`min`,`max`,`avg`,`device_id`,`timestamp`)  VALUES ('3','3','0','0.52','0.74','0.60','13',NOW())] 
SNMP Check response code: 0
SNMP[/usr/bin/snmpget -v3 -n '' -l 'authPriv' -a 'SHA' -A 'librenms' -u 'librenms' -x 'AES' -X 'librenms' -OUQn -M /opt/librenms/mibs udp:HOSTNAME:161 SNMPv2-MIB::sysDescr.0 SNMPv2-MIB::sysObjectID.0]
.*.*.0 = Allied Telesis router/switch, AW+ v5.4.7-2.1
.*.*.0 = .*.*.1.82

| Allied Telesis router/switch, AW+ v5.4.7-2.1 | .1.3.6.1.4.1.207.1.1.82 | 

Device os was updated to alliedtelesis!SQL[UPDATE `devices` set `os` ='alliedtelesis' WHERE `device_id` = '13'] 
SQL[INSERT INTO `eventlog` (`host`,`device_id`,`reference`,`type`,`datetime`,`severity`,`message`,`username`)  VALUES ('13','13','NULL','system',NOW(),'3','Device type changed server => ','')] 
SQL[UPDATE `devices` set `type` ='' WHERE device_id='13'] 
Device type changed to !
SQL[SELECT attrib_value FROM devices_attribs WHERE `device_id` = '13' AND `attrib_type` = 'poll_mib' ] 

Modules status: Global+ OS+ Device  
#### Load disco module os ####
SNMP[/usr/bin/snmpget -v3 -n '' -l 'authPriv' -a 'SHA' -A 'librenms' -u 'librenms' -x 'AES' -X 'librenms' -OUQn -M /opt/librenms/mibs:/opt/librenms/mibs/alliedtelesis udp:HOSTNAME:161 SNMPv2-MIB::sysDescr.0 SNMPv2-MIB::sysObjectID.0]
.*.*.0 = Allied Telesis router/switch, AW+ v5.4.7-2.1
.*.*.0 = .*.*.1.82

| Allied Telesis router/switch, AW+ v5.4.7-2.1 | .1.3.6.1.4.1.207.1.1.82 | 
SNMP[/usr/bin/snmpget -v3 -n '' -l 'authPriv' -a 'SHA' -A 'librenms' -u 'librenms' -x 'AES' -X 'librenms' -Oqv -M /opt/librenms/mibs:/opt/librenms/mibs/alliedtelesis udp:HOSTNAME:161 .1.3.6.1.4.1.12148.10.2.6.0]
No Such Object available on this agent at this OID

SQL[INSERT INTO `eventlog` (`host`,`device_id`,`reference`,`type`,`datetime`,`severity`,`message`,`username`)  VALUES ('13','13','NULL','system',NOW(),'3','Device OS changed alliedtelesis => generic','')] 
SQL[UPDATE `devices` set `os` ='generic' WHERE device_id='13'] 
Changed OS: Generic Device (generic)

```

I've tried changing back to the original allied.inc.php but still get the same issue...

Thanks,

Matt

laf commented:
It's hard to see from that as the debug output masks some info. Run `./discovery.php -h c3-vpn -v -m os` - check the output doesn't contain sensitive info and post it up.

Sometimes the os cache can get in the way so run `rm -f cache/os_defs.cache` first

murrant commented:
@mattie47 Your discovery file needs to match the os name `alliedwareplus.yaml` -> `alliedtelesis.yaml`

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/8f81d85f-d4c0-42f0-913c-266a094f4cb4) completed: **2 updated code elements**

mattie47 commented:
@murrant  

Ahhhhh thanks! 

I was looking over all the files with a fine toothed comb but hadn't clicked that the file name had changed...I previously had it as alliedwareplus, but changed it to alliedtelesis later....

I must have stuffed something up earlier on doing a git pull or something. 

Everything is working nicely now!

![image](https://user-images.githubusercontent.com/4696925/32630708-8a35dc02-c602-11e7-9c8b-188a1f7b3560.png)

If I have time today, I'll have a look at some other non-AW+ Allied Telesis platforms to confirm behaviour. 

Thanks,

Matt

laf commented on includes/definitions/alliedtelesis.yaml:
```diff
@@ -1,10 +1,14 @@
-os: allied
-text: AlliedWare
+os: alliedtelesis
+text: 'AW+'
```
Is this descriptive enough?

laf commented on includes/definitions/alliedtelesis.yaml:
```diff
@@ -1,10 +1,14 @@
     -
       sysObjectId: .1.3.6.1.4.1.207.
       sysObjectId_except: .1.3.6.1.4.1.207.1.4.126
+    - sysDescr: 'Allied Telesis'
```
Is this needed. You test file uses `.1.3.6.1.4.1.207.1.14.137` which would be picked up by the original check.

laf commented on includes/definitions/alliedtelesis.yaml:
```diff
@@ -1,10 +1,14 @@
-os: allied
-text: AlliedWare
+os: alliedtelesis
```
is aw+ the os/firmware name? If so, the actual OS for us should be that - maybe `awplus`

laf commented on includes/polling/os/alliedtelesis.inc.php:
```diff
@@ -1,21 +1,39 @@
+
+//Have removed above syntax and tried to make below work.
+
+$hardware = trim(snmp_getnext($device, "rscBoardName", "-OQv", "AT-RESOURCE-MIB"), '"');
```
What happens for older hardware, will these return any data? If not, you should do a check for empty variables and fall back to the old snmp_get calls.

laf commented on mibs/alliedtelesis/at-almmon.mib:
```diff
@@ -0,0 +1,253 @@
+--  ============================================================================
+-- AT-ALMMON-MIB, Allied Telesis enterprise MIB:
```
This mib file needs changing to `mibs/alliedtelesis/AT-ALMMON-MIB`

laf commented on mibs/alliedtelesis/at-atmf.mib:
```diff
@@ -0,0 +1,608 @@
+-- ============================================================================
```
This mib filename needs changing to `mibs/alliedtelesis/AT-ATMF-MIB`

laf commented on mibs/alliedtelesis/at-boards.mib:
```diff
@@ -0,0 +1,1287 @@
+-- ============================================================================
```
This mib filename needs updating to `mibs/alliedtelesis/AT-BOARDS-MIB`

laf commented on mibs/alliedtelesis/at-chassis.mib:
```diff
@@ -0,0 +1,318 @@
+-- ============================================================================
```
This mib filename needs updating to `mibs/alliedtelesis/AT-CHASSIS-MIB`

laf commented on mibs/alliedtelesis/at-dhcpsn.mib:
```diff
@@ -0,0 +1,347 @@
+-- ============================================================================
```
This mib filename needs updating to `mibs/alliedtelesis/AT-DHCPSN-MIB`

laf commented on mibs/alliedtelesis/at-dns.mib:
```diff
@@ -0,0 +1,189 @@
+-- ============================================================================
```
This mib filename needs updating to `mibs/alliedtelesis/AT-DNS-CLIENT-MIB`

laf commented on tests/snmpsim/allied.snmprec:
```diff
@@ -1,2 +0,0 @@
-1.3.6.1.2.1.1.1.0|4|AT-GS950/24
```
You should leave this check in place and just rename it to `tests/snmpsim/alliedtelesis_gs950.snmprec`

mattie47 commented:
Hi @laf,

Thanks for the review comments. I'll try and get to those over the next couple of days.

So now things are working for devices running Alliedware Plus, however as you've obviously identified where does this leave support for legacy devices? 

This is what I'm currently thinking about and wondering how to solve. My train of thought now is I perhaps should be doing two yaml files, one being allied (and essentially covers anything legacy), and one called awplus (for newer alliedware Plus products). 

I currently have 3 or 4 older products I'm going to try get SNMP set up on and check behaviour.

In short, the enterprise vendor OID is the same (207), and I'm currently using this as the discovery identifier:

```
 discovery:
     -
       sysObjectId: .1.3.6.1.4.1.207.
```

Given this, how do you think I should be distinguishing between older products running alliedware, and newer products running awplus? 

The primary distinguisher I can think of is the sysObjectId, but can you think of a better way of doing this?

E.g.

```
// Device examples pre-2010
// sysDescr.0 = STRING: "Allied Telesis AT-8624T/2M version 2.9.1-13 11-Dec-2007"
// sysDescr.0 = STRING: "Allied Telesyn Ethernet Switch AT-8012M"
// sysDescr.0 = STRING: "ATI AT-8000S" <------------------------------------- RADLAN ********
// sysDescr.0 = STRING: "Allied Telesyn AT-8624T/2M version 2.8.1-02 05-Sep-2006"
// sysDescr.0 = STRING: "AT-8126XL, AT-S21 version 1.4.2"
// AtiL2-MIB::atiL2SwProduct.0 = STRING: "AT-8326GB"
// AtiL2-MIB::atiL2SwVersion.0 = STRING: "AT-S41 v1.1.6 "


// Almost all current Allied Telesis switches and routers run AW+. This is a current example:
// sysDescr.0 = STRING: Allied Telesis router/switch, AW+ v5.4.7-2.1

``


Thanks,

Matt

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/6b3c1ed8-e4f0-4d02-89be-625d66a702fb) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/6b3c1ed8-e4f0-4d02-89be-625d66a702fb) completed: **No new issues**

laf commented:
> Given this, how do you think I should be distinguishing between older products running alliedware, and newer products running awplus?
> 
> The primary distinguisher I can think of is the sysObjectId, but can you think of a better way of doing this?

That depends on what the data shows :) If the older devices do have a different OID - even if it's just the last octet then we can include those specific ones in the old allied OS leaving awplus to pick up the rest.

mattie47 commented:
> That depends on what the data shows :) If the older devices do have a different OID - even if it's just the last octet then we can include those specific ones in the old allied OS leaving awplus to pick up the rest.

OK no worries. I've got a couple older boxes in LibreNMS now so will see what I can figure out today...

Thanks,

Matt

murrant commented:
You can also do a combination of sysObjectID and sysDescr if needed.  Assuming all AliedWare Plus sysDescr contain the `AW+` string
AW:
```yaml
discovery:
    -
      sysObjectId: .1.3.6.1.4.1.207.
      sysDescr_except: 'AW+'
```
AW+:
```yaml
discovery:
    -
      sysObjectId: .1.3.6.1.4.1.207.
      sysDescr: 'AW+'
```

mattie47 commented:
@murrant 

Ah! That's a great idea. Will try this.

Thanks :-)

mattie47 commented:
>Sometimes the os cache can get in the way so run rm -f cache/os_defs.cache first

This would be really useful on the "Initial-Detection" docs page. Especially when you're making generic changes in your configs trying to get things to work. 

I suddenly remembered laf had said this after an hour or so of trying to figure out why I couldn't get the older products in a different yaml config file working. 


I've separated configs between Alliedware Plus and other allied OSs now.  Still got a bit more testing and getting thing working for the other OSes before I'll do another push.

Thanks,

Matt

laf commented:
> This would be really useful on the "Initial-Detection" docs page. Especially when you're making generic changes in your configs trying to get things to work.

Edit the docs in a new PR :)

> I've separated configs between Alliedware Plus and other allied OSs now. Still got a bit more testing and getting thing working for the other OSes before I'll do another push.

Looking forward to it.

mattie47 commented:
Hi laf/murrant ,


So I've done some further digging and analysis into the different Allied Telesis products.

Sorry this post is probabably a bit long....

Basically there seem to be 4 main operating systems. 3 of which are pretty much EOL but have made up a number of the devices in murrant's previous report.


**I've done the following:**

- Aquired the latest available mibs for each OS. 
- Done SNMP walks on devices running each OS. 
- Worked out relevant OID for each LibreNMS variable such as vendor/version.


**Some Context:**

- **Radlan** - Allied Telesis products such as 8000S and 8000GS seem to run this. This is confirmed by downloading the latest official mibs mentioning radlan in them. As previously mentioned, this OS is on mass-produce low cost switches by Marvell (Some Allied Telesis, some not). 
- **Alliedware Plus version 5.x.x** is the main OS of all current decent switches/routers.
- **Alliedware Plus version 2.x.x** is an OS which shares the same name as 5.x.x above however it's really not the same product. CLI feels the same is only found on mostly EOL L2/Edge products. 
- **Alliedware OS** is EOL and has been for some time. This was the previous main OS for premium devices. 




So regarding the table of devices, **I've changed the OS name** to more correctly reflect what they are:

User contributed stats
--
Total | OS | sysObjectID | sysDescr
-- | -- | -- | --
25 | unknown | .1.3.6.1.4.1.207.1.4.167 | AT-GS950/24 Gigabit Ethernet WebSmart Switch
25 | radlan | .1.3.6.1.4.1.207.1.4.127 | ATI AT-8000S
24 | radlan | .1.3.6.1.4.1.207.1.4.128 | ATI AT-8000S
13 | radlan | .1.3.6.1.4.1.207.1.4.145 | 48-port 10/100/1000 Ethernet Switch
12 | radlan | .1.3.6.1.4.1.207.1.4.144 | 24-port 10/100/1000 Ethernet Switch with PoE
9 | awplusv5 | .1.3.6.1.4.1.207.1.4.198 | Allied Telesis router/switch, AW v5.4.5-2.2
8 | awplusv2 | .1.3.6.1.4.1.207.1.4.237 | AlliedWare Plus (TM) 2.4.1.0
3 | awplusv2 | .1.3.6.1.4.1.207.1.4.151 | AlliedWare Plus (TM) 2.1.8.0
4 | alliedware | .1.3.6.1.4.1.207.1.14.53 | Allied Telesyn AT-9924T-EMC version 2.6.6-01 15-Apr-2005
3 | alliedware | .1.3.6.1.4.1.207.1.14.49 | Allied Telesis AT-9924T version 2.9.2-14 24-Jun-2014


Unfortunately I don't have one of the AT-GS950s around to double check things but I've tried all the others. 

# Radlan

```
// sysDescr.0 = STRING: "ATI AT-8000S"
// sysDescr.0 = STRING: 48-port 10/100/1000 Ethernet Switch 
// sysDescr.0 = STRING: 24-port 10/100/1000 Ethernet Switch with PoE
```

Radlan - 48-port 10/100/1000 Ethernet Switch aka 8000GS.
--
Variable | snmpwalk | Response |
-- | -- | -- 
System Description | snmpget -v2c -c librenms 10.37.105.80 sysDescr.0            | SNMPv2-MIB::sysDescr.0 = STRING: 48-port 10/100/1000 Ethernet Switch 
Hostname           | snmpget -v2c -c librenms 10.37.105.80 sysName.0             | SNMPv2-MIB::sysName.0 = STRING: switch-8000GS
Contact            | snmpget -v2c -c librenms 10.37.105.80 sysContact.0          | SNMPv2-MIB::sysContact.0 = STRING: matt
Location           | snmpget -v2c -c librenms 10.37.105.80 sysLocation.0         | SNMPv2-MIB::sysLocation.0 = STRING: lab
Version            | snmpget -v2c -c librenms 10.37.105.80 .1.3.6.1.4.1.89.2.4.0 | SNMPv2-SMI::enterprises.89.2.4.0 = STRING: "1.0.0.33" 
Device Model	   | snmpget -v2c -c librenms 10.37.105.80 sysObjectID.0  		 | SNMPv2-MIB::sysObjectID.0 = OID: AT-PRODUCT-MIB::swhub.145
Alternative Device. Not sure if this will always work. Need to verify on 8000S   | snmpgetnext -On -v2c -c librenms 10.37.105.80 .1.3.6.1.2.1.47.1.1.1.1.2.64  | ENTITY-MIB::entPhysicalDescr.67108992 = STRING: AT-8000GS/48
Serial Number	   | snmpgetnext -v2c -c librenms 10.37.105.80 ENTITY-MIB::entPhysicalSerialNum.64 | ENTITY-MIB::entPhysicalSerialNum.67108992 = STRING: A03970G084800039 C


## Alliedware Plus 5.x.x:

### General list of products:

```
Switches: x200, x210, x230, x310, x510, x530, x600, x610, x900, x930
Chassis: SBx8100, SBx908, SBx908GEN2
Routers: AR2010V, AR2050, AR3050S, AR4050S
```

The yaml file needs to match on both AW+ and/or "AW v5.4.x". So while matching on AW+ works fine on current releases, it seems releases from 3+ years ago used "AW v5.4.x" which can be observed in the user contributed stats above. 

Alternatively I guess "Allied Telesis router/switch" would actually work because this descr has been the same for all 5.4.x releases. 

Alliedware Plus 5.x.x:
--
Variable 		   | snmpget | Response |
-- | -- | -- 
System Description | snmpget -v2c -c librenms 10.37.105.11 sysDescr.0          			    	| SNMPv2-MIB::sysDescr.0 = STRING: Allied Telesis router/switch, AW+ 5.4.7-2.1
Hostname           | snmpget -v2c -c librenms 10.37.105.11 sysName.0                            | SNMPv2-MIB::sysName.0 = STRING: c2-core-stk
Contact            | snmpget -v2c -c librenms 10.37.105.11 sysContact.0        			    	| SNMPv2-MIB::sysContact.0 = STRING: Matt
Location           | snmpget -v2c -c librenms 10.37.105.11 sysLocation.0        			    | SNMPv2-MIB::sysLocation.0 = STRING: Campus 1
Version            | snmpget -On -v2c -c librenms 10.37.105.11 currSoftVersion.0				| AT-SETUP-MIB .1.3.6.1.4.1.207.8.4.4.4.500.2.1.1.0 = STRING: 5.4.7-2.1
Device Model	   | snmpgetnext -On -v2c -c librenms 10.37.105.11 rscBoardName					| AT-RESOURCE-MIB .1.3.6.1.4.1.207.8.4.4.3.21.1.1.4.1.3 = STRING: AT-x930-52GTX
Serial Number	   | snmpgetnext -On -v2c -c librenms 10.37.105.11 rscBoardSerialNumber 		| AT-RESOURCE-MIB::rscBoardSerialNumber.1.3 = STRING: A04878A143000034


## Alliedware Plus 2.x.x

### General list of products:

```
-- AT-9000
-- AT-8100S
-- AT-8100L
-- AT-FS970M
```

```
// sysDescr.0 = STRING: AlliedWare Plus (TM) 2.1.8.0
```


Alliedware Plus 2.x.x
--
Variable 		   | snmpwalk 												   					| Response |
-- | -- | -- 
System Description | snmpget -v2c -c librenms 10.37.105.85 sysDescr           			    	| SNMPv2-MIB::sysDescr.0 			 = STRING: AlliedWare Plus (TM) 2.2.3.0
Hostname           | snmpget -v2c -c librenms 10.37.105.85 sysname                             	| SNMPv2-MIB::sysName.0 			 = STRING: sw-x8100s
Contact            | snmpget -v2c -c librenms 10.37.105.85 syscontact          			    	| SNMPv2-MIB::sysContact.0 			 = STRING: matt
Location           | snmpget -v2c -c librenms 10.37.105.85 syslocation        			    	| SNMPv2-MIB::sysLocation.0			 = STRING: lab
Version            | snmpget -v2c -c librenms 10.37.105.85 .1.3.6.1.4.1.207.8.17.1.3.1.5.1 		| AT-SMI-MIB::mibObject.17.1.3.1.5.1 = STRING: "2.2.3.0" 
Device Model	   | snmpget -v2c -c librenms 10.37.105.85 .1.3.6.1.4.1.207.8.17.1.3.1.6.1		| AT-SMI-MIB::mibObject.17.1.3.1.6.1 = STRING: "AT-8100S/48POE"
Serial Number	   | snmpget -v2c -c librenms 10.37.105.85 .1.3.6.1.4.1.207.8.17.1.3.1.8.1	    | AT-SMI-MIB::mibObject.17.1.3.1.8.1 = STRING: "A04481G114700071"


## Alliedware

### General List of Alliedware Products:

```
Routers:
AR410
AR720, 725, 740, 745 
AR415, 440-1-2, 450	
AR550, 560, 570, 750

Switches:
AT-8624T, AT-8648 
AT-8824, 8848 	
Rapier 24, 24i, 48, 48i (CenterCom (86/87)(24/48)XL)
AT-8724, 8748 
AT-8948, 8948i 
AT-9924SP, 9924T
AT-9812, 9816 
Switchblade 4000
```

```
// sysDescr.0 = STRING: "Allied Telesis AT-8624T/2M version 2.9.1-13 11-Dec-2007"
// sysDescr.0 = STRING: "Allied Telesyn AT-8624T/2M version 2.8.1-02 05-Sep-2006"
```

It seems sysXXX.0 e.g. sysDescr.0 is required unlike AWPlusv2 above. However explicitely using xxx.0 also works for AWPlusv2.

Alliedware:
--
Variable 		   | snmpwalk 												   					| Response |
-- | -- | -- 
System Description | snmpget -v2c -c librenms 10.37.105.81 sysDescr.0          			    	| SNMPv2-MIB::sysDescr.0 			 = STRING: CentreCOM 9924Ts, version 3.2.1-04, built 08-Sep-2009
Hostname           | snmpget -v2c -c librenms 10.37.105.81 sysName.0                            | SNMPv2-MIB::sysName.0 		 	 = STRING: SNMPv2-MIB::sysName.0 = STRING: aw-switch
Contact            | snmpget -v2c -c librenms 10.37.105.81 sysContact.0        			    	| SNMPv2-MIB::sysContact.0 			 = STRING: matt
Location           | snmpget -v2c -c librenms 10.37.105.81 sysLocation.0        			    | SNMPv2-MIB::sysLocation.0 	 	 = STRING: Lab1
Version            | Not Possible via SNMP. Get from SysDesc.0 									| version 3.2.1-04
Device Model	   | snmpget -On -v2c -c librenms 10.37.105.81 .1.3.6.1.4.1.207.8.4.4.5.2.1.3.1	| AT-SMI-MIB::arInterfaces.2.1.3.1 	 = STRING: "CentreCOM 9924Ts"
Serial Number	   | snmpget -On -v2c -c librenms 10.37.105.81 .1.3.6.1.4.1.207.8.4.4.5.2.1.5.1 | AT-SMI-MIB::arInterfaces.2.1.5.1   = STRING: "45AX49005"



So - I've got all the relevant information for the OSes above. 

# Mibs

What exactly does LibreNMS want in terms of MIBs? All the mibs the device has? Or simply the ones required for OS Discovery?

The mibs I have are the latest available and all to some degree are newer than what's in LibreNMS. 

**To summarise:**
radlan - Roughly 56 mib files
awplusv5 - 45 mib files
awplusv2 - 4 mib files 
alliedware - 31 mibs. There's some overlap with the awplusv5 mibs, but I haven't looked in detail as to where. 

Does LibreNMS handle mibs stored within subfolders of the mib directory? 
i.e. mibs/alliedtelesis/awplusv5 mibs/alliedtelesis/awplusv2 mibs/alliedtelesis/alliedware/ 


# This is where I ask for help

I was hoping either of you guys (or someone else on the LibreNMS team) might be able to assist with writing the inc.php file for the legacy "allied" devices, given the information I've presented above. 

I'm not a software dev, and I just don't have time to be trying to learn PHP for this :-/

I can certainly test anything and more than happy to do this. I just don't know how to approach the logic of above into a conf file. 


**What I'm asking for:**
- allied.incl.php to include the various variables for legacy OSes awplusv2 and alliedware. 
- Radlan - I've worked out what the mib objects are to get the various variables, but I don't know if this would belong in allied.inc.php, or in radlan.inc.php which already exists. 

I included awplusv5 above in the output for completeness, however given I already have this working I don't consider anything necessary here.

You'll probably notice in some places I've included just the mib object name e.g. rscBoardName, however in output I've ensured I've included the full OID and relevant MIB file. 


Thoughts?


Thanks!

Matt

murrant commented:
@mattie47 If you can provide us with snmpwalks (based on https://docs.librenms.org/#Support/FAQ/#what-information-do-you-need-to-add-a-new-os )  we can simulate the devices and try things out / tweak them when we have time.

mattie47 commented:
Hi @murrant,

OK thanks.

I'll do this for each device type when I have a chance. 

Giving this PR some more thought over the weekend, I've come to the conclusion I really should be submitting it in 2 phases:

**Phase 1 - Simply get AW+ support in libreNMS, along with the AW+ MIBs, and SVGs etc.** 

This is the crux of this PR, along with ensuring that legacy AT devices still work to the same current level of support.

At this point I pretty much have things working nicely like this, but need to double check a few things before sending final PR commits, along with fixing any concerns previously made by @laf. 

**Phase 2 - Improve Legacy Allied Telesis/Telesyn product support.** 

This would be improving the allied.inc.php file to include extra details such as Device/Hostname/Serial/SW Version etc for each device type. These details are stated in my previous comment. 

So - I'll work towards getting this finished for phase 1. 


- Would you mind having a look at my questions under MIBs above? Reading through the Libre docs, I'm just not sure what's actually desired by the project when it comes to vendor mibs. 


Thanks,

Matt

murrant commented:
@mattie47 Sounds great.

As for the mibs.  We only support one level of sub directories under the mibs directory.

By default we add the parent mib directory and the sub-directory with the same name as the OS.  An OS can specify an alternate mib directory (usually if multiple os share one directory)

You can probably put all the mibs under one directory unless there are conflicting mibs.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/5358c42e-7c9a-491a-8097-194d3165aa99) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/5358c42e-7c9a-491a-8097-194d3165aa99) completed: **No new issues**

mattie47 commented:
@murrant  Thanks for that.

I've split the mibs into allied (for anything legacy and general), and awplus.

This is also because there are some conflicts between the mibs, and it's just not worth the time trying to figure out where these are.

The conflicts I refer to are when I was doing snmpwalks on the older products with the older mibs verses the newer mibs. Some files are the same mib name, but for legacy products I only got detail in the old mibs.


Anyway - I've submitted my changes now as one commit, and I also addressed any concerns @laf  previously had in this.

Thanks :-)

Matt

murrant commented on includes/polling/os/awplus.inc.php:
```diff
@@ -0,0 +1,12 @@
+//$hardware and $serial use snmp_getnext as the OID for these is not always fixed.
+//However, the first OID is the device baseboard. 
+
+$hardware = trim(snmp_getnext($device, "rscBoardName", "-OQv", "AT-RESOURCE-MIB"), '"');
```
The trims here should not be needed snmp_get and snmp_getnext trim surrounding "

mattie47 replied:
Done :-)

murrant commented:
After removing those extra trims, looks good to merge for me :)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/60785396-bf84-48b9-a62e-464c37622fb7) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/60785396-bf84-48b9-a62e-464c37622fb7) completed: **No new issues**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/5d6e1f77-ffa5-4445-8ed2-40d799bd7981) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/5d6e1f77-ffa5-4445-8ed2-40d799bd7981) completed: **No new issues**

laf commented on html/images/os/allied.svg:
```diff
@@ -0,0 +1 @@
+alliedtelesis.svg
```
Looks like the svg has gone wrong.

mattie47 replied:
Nah - this is a soft symlink to alliedtelesis.svg. 

I saw no reason to duplicate the svg file for multiple OSes, so used a symlink to do this.

laf replied:
Then just have on image and change the icon the yaml discovery file to point to that one image :)

mattie47 replied:
Ah yeah fair comment.

laf commented on includes/definitions/allied.yaml:
```diff
@@ -1,10 +1,16 @@
 ifname: 1
 over:
     - { graph: device_bits, text: Traffic }
+#exclude ...207.1.4.126 for ATI 8000S/8000GS which identify as "radlan" OS. 
```
You can put this next to line it's intended for with a # to show it's a comment or just remove it.

mattie47 replied:
K thanks.

laf commented on includes/definitions/awplus.yaml:
```diff
@@ -0,0 +1,19 @@
+ifname: 1
+over:
+    - { graph: device_bits, text: Traffic }
+#exclude ...207.1.4.126 for ATI 8000S/8000GS which identify as "radlan" OS. 
```
Same for this.

mattie47 replied:
Will do this in next commit

laf commented on includes/definitions/radlan.yaml:
```diff
@@ -8,3 +8,5 @@ over:
 discovery:
     - sysDescr:
         - AT-8000
+        - '48-port 10/100/1000 Ethernet Switch'
```
This seems too generic to me. How come this is needed?

mattie47 replied:
Yeah, it's a bit rubbish.  The system description is literally "48-port 10/100/1000 Ethernet Switch". Doesn't say vendor or model etc. 

Was this the correct way of specifying multiple SysDescr to potentially match on? I was trying to do "AT-8000" OR "48-port 10/100/1000 Ethernet Switch".

laf replied:
Better to use sysObjectID if it has a unique one?

mattie47 replied:
Fair point. 

I'll just remove the line from this commit anyway.

As above, how do you go between "This AND that", and "This OR that" in Yaml ?

laf replied:
You have OR right now with these two lines:

```
        - AT-8000
        - '48-port 10/100/1000 Ethernet Switch'
```

Basically meaning either.

For and:

```
    - sysObjectId:
        - .1.3.6.1.4.1.1129.1.2.
    - sysDescr:
        - TOSHIBA e-STUDIO
```

This requires both of those to match.

laf commented on includes/polling/os/awplus.inc.php:
```diff
@@ -0,0 +1,12 @@
+<?php
+
+//OS info - https://www.alliedtelesis.com/products/software/AlliedWare-Plus
```
You can remove these comments.

mattie47 replied:
Do you mean the next line specifying OS datasheet also? 

Is it not desirable to have any comments about the OS in the config file?

laf replied:
You've got them already in the discovery file. Don't see a need to duplicate them.

mattie47 replied:
Ah yes - True...

laf commented on mibs/awplus/mib_info.txt:
```diff
@@ -0,0 +1,5 @@
+This folder contains MIBs for Allied Telesis products running AW+ 5.x.x
```
imho you can get rid of this file.

mattie47 replied:
OK - no problem.

I created it to try clear up why some mibs have the same names between allied folder, and awplus. It's because despite the names being the same, the content of said mibs differ slightly.

I just thought someone might come along later and go "why are there duplicate files here?".  You probably noted I put more detail in the txt file in the "allied" mibs folder also.

laf replied:
I'm not sure it's needed. Anything in a sub folder is in essence after the mib dir revamp. Everything before was in one folder, now we are picky about how they are done.

mattie47 replied:
kk

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/30f96f95-24e0-4c4f-8f87-2203d2c62e8e) was created.

laf commented:
All good from me. @librenms/reviewers?

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/30f96f95-24e0-4c4f-8f87-2203d2c62e8e) completed: **No new issues**

murrant commented:
LGTM
